src/subdomains/staking/application/services/staking.service.ts

```diff
@@ -149,4 +179,96 @@ export class StakingService {
+
+    // not querying Stakings, because eager query is not supported, thus unsafe to fetch entire entity
+    const queryByWithdrawals = await this.repository
+      .createQueryBuilder('staking')
+      .innerJoin('staking.withdrawals', 'withdrawal')
+      .where('withdrawal.status = :withdrawalStatus', { withdrawalStatus: WithdrawalStatus.CONFIRMED })
+      .andWhere('(withdrawal.amountEur IS NULL OR withdrawal.amountUsd IS NULL OR withdrawal.amountChf IS NULL)')
+      .leftJoinAndSelect('staking.asset', 'asset')
+      .getMany()
+      .then((s) => s.map((i) => ({ stakingId: i.id, assetId: i.asset.id })));
+
+    const stakingReferences = this.removeStakingReferencesDuplicates(queryByDeposits.concat(queryByWithdrawals));
+
+    stakingReferences.length > 0 &&
+      console.info(
+        `Adding fiat references to ${stakingReferences.length} staking(s). Staking Id(s):`,
+        stakingReferences.map((s) => s.stakingId),
+      );
+
+    return stakingReferences;
+  }
+
+  private async getReferencePrices(stakings: StakingReference[]): Promise<Price[]> {
+    const prices = [];
+    const uniqueAssetIds = [...new Set(stakings.map((s) => s.assetId))];
+
+    for (const assetId of uniqueAssetIds) {
+      for (const fiatName of Object.values(Fiat)) {
+        try {
+          const price = await this.fiatPriceProvider.getFiatPrice(fiatName, assetId);
+
+          prices.push(price);
+        } catch (e) {
+          console.error(`Could not find fiat price for assetId ${assetId} and fiat '${fiatName}'`, e);
+          continue;
+        }
+      }
+    }
+
+    return prices;
+  }
+
+  private async calculateFiatReferencesForStakings(stakings: StakingReference[], prices: Price[]): Promise<void> {
+    const confirmedStakings = [];
+
+    for (const ref of stakings) {
+      try {
+        await this.calculateFiatReferencesForStaking(ref.stakingId, prices);
+        confirmedStakings.push(ref.stakingId);
+      } catch (e) {
+        console.error(
+          `Could not calculate fiat reference amount for Staking Id: ${ref.stakingId}. Asset Id: ${ref.assetId}`,
+          e,
+        );
+        continue;
+      }
+    }
+
+    confirmedStakings.length > 0 &&
+      console.info(
+        `Successfully added fiat references to ${confirmedStakings.length} staking(s). Staking Id(s):`,
+        confirmedStakings,
+      );
+  }
+
+  private async calculateFiatReferencesForStaking(stakingId: number, prices: Price[]): Promise<void> {
+    const staking = await this.repository.findOne(stakingId);
+
+    staking.calculateFiatReferences(prices);
+    await this.repository.save(staking);
+  }
+
+  private removeStakingReferencesDuplicates(stakings: StakingReference[] = []): StakingReference[] {
+    return stakings.reduce((res, curr) => {
+      const existing = res.find((r) => r.stakingId === curr.stakingId);
+
+      !existing && res.push(curr);
+
+      return res;
+    }, []);
```

Comment:
It should be possible to remove duplicates using
```ts
stakings.filter((item, pos, self) => self.indexOf(item) === pos);
```

Reply:
addressed